### PR TITLE
BREAKING(async): deprecate `deferred()` in favor of `Promise.withResolvers()`

### LINKED ISSUE #3757: Deprecate "deferred" API
With Deno v1.38 a new [`Promise.withResolvers()`](https://github.com/tc39/proposal-promise-with-resolvers) API will be available which has the same purpose as `deferred`. We should deprecate the latter and guide users to use the new built-in API.

### PR DESCRIPTION
**Pull request by lino-levan**
closes #3757

### CODE REVIEW

File: async/deferred.ts [hunk 1]
```diff
@@ -6,6 +6,8 @@
 // typescript produces broken code when targeting ES5 code.
 // See https://github.com/Microsoft/TypeScript/issues/15202
 // At the time of writing, the github issue is closed but the problem remains.
+
+/** @deprecated (will be removed after 0.214.0) */
```
Comment (lucacasonato) [inline async/deferred.ts:10]: Add the same instruction to use `Promise.withResolvers` here.

Comment (lino-levan) [reply async/deferred.ts:10]: This is a type? Do we still want to do that?

File: async/deferred.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@
 // typescript produces broken code when targeting ES5 code.
 // See https://github.com/Microsoft/TypeScript/issues/15202
 // At the time of writing, the github issue is closed but the problem remains.
+
+/** @deprecated (will be removed after 0.214.0) Use `Promise.withResolvers` instead. */
```
Comment (iuioiua) [inline async/deferred.ts:10]:
```suggestion
/** @deprecated (will be removed after 0.207.0) Use {@linkcode Promise.withResolvers} instead. */
```

File: async/deferred.ts [hunk 3]
```diff
@@ -14,6 +16,8 @@ export interface Deferred<T> extends Promise<T> {
 }
 
 /**
+ * @deprecated (will be removed after 0.214.0) Use `Promise.withResolvers` instead.
```
Comment (iuioiua) [inline async/deferred.ts:19]:
```suggestion
 * @deprecated (will be removed after 0.207.0) Use {@linkcode Promise.withResolvers} instead.
```

**Review by iuioiua — CHANGES_REQUESTED** [review]
We'll confirm this removal version before merging. But it's also good to link to docs like this going forward. ~~This should fail on stable but pass on canary.~~

File: async/deferred.ts [hunk 4]
```diff
@@ -6,6 +6,8 @@
 // typescript produces broken code when targeting ES5 code.
 // See https://github.com/Microsoft/TypeScript/issues/15202
 // At the time of writing, the github issue is closed but the problem remains.
+
+/** @deprecated (will be removed after 0.207.0) Use {@linkcode Promise.withResolvers} instead. */
```
Comment (iuioiua) [inline async/deferred.ts:10]:
```suggestion
/** @deprecated (will be removed in 0.208.0) Use {@linkcode Promise.withResolvers} instead. */
```

File: async/deferred.ts [hunk 5]
```diff
@@ -14,6 +16,8 @@ export interface Deferred<T> extends Promise<T> {
 }
 
 /**
+ * @deprecated (will be removed after 0.207.0) Use {@linkcode Promise.withResolvers} instead.
```
Comment (iuioiua) [inline async/deferred.ts:19]:
```suggestion
 * @deprecated (will be removed in 0.208.0) Use {@linkcode Promise.withResolvers} instead.
```

**Review by iuioiua — CHANGES_REQUESTED** [review]
Yoshiya and I agreed that a deprecation period of 2 minor versions (12 weeks) is reasonable. Let's go with that.

**Review by iuioiua — APPROVED** [review]
LGTM. Thanks, Lino!

File: async/deferred.ts [hunk 6]
```diff
@@ -6,6 +6,8 @@
 // typescript produces broken code when targeting ES5 code.
 // See https://github.com/Microsoft/TypeScript/issues/15202
 // At the time of writing, the github issue is closed but the problem remains.
+
+/** @deprecated (will be removed in 0.208.0) Use {@linkcode Promise.withResolvers} instead. */
```
Comment (iuioiua) [inline async/deferred.ts:10]:
```suggestion
/** @deprecated (will be removed in 0.207.0) Use {@linkcode Promise.withResolvers} instead. */
```

File: async/deferred.ts [hunk 7]
```diff
@@ -14,6 +16,8 @@ export interface Deferred<T> extends Promise<T> {
 }
 
 /**
+ * @deprecated (will be removed in 0.208.0) Use {@linkcode Promise.withResolvers} instead.
```
Comment (iuioiua) [inline async/deferred.ts:19]:
```suggestion
 * @deprecated (will be removed in 0.207.0) Use {@linkcode Promise.withResolvers} instead.
```

**Review by iuioiua — APPROVED** [review]
I spoke with Yoshiya again, and we'll reduce the deprecation period for this to 1 minor version (6 weeks). We'll likely merge this PR right before the next minor release. LGTM again and thanks again!

**Review by iuioiua — APPROVED** [review]
Sorry again for the change. I previously misunderstood specifics around the deprecation period. Either way, I've now changed this to be removed in 0.209.0 and can now merge. Thanks again for this, Lino!

**Pull request merged**